README.txt
###########################################################
INSTALLATION.

	Make sure your hosting service meets the requirements.
	https://milliondollarscript.com/article/hosting/

	You may find these online installation instructions useful:
	https://milliondollarscript.com/article/docs/
	
	More help and support is available at the forum.
	https://milliondollarscript.com/forums/

	You can also join the official Discord server.
	https://discord.gg/9AjyZb6

	Knowledge Base / FAQ
	https://milliondollarscript.com/category/faq/

	Note: The documentation on the website is currently undergoing a rewrite. Sorry for the mess!

	If you'd like me (the developer) to install it for you then you can order the install service here:
	https://milliondollarscript.com/product/million-dollar-script-installation-service/

#############################################################

1. Setup your database.

	Create a new database on your hosting account.
	(Make sure that your MYSQL database user has privileges to INSERT, UPDATE,
	REPLACE and LOCK TABLES.)
	
	If you need detailed instructions for setting up a MySQL database, then 
	these can be found in the manual / docs of your web hosting provider.
	
	Here is an overview of the steps:
	 a. Create a new database
	 b. Create a new database user
	 c. Add the new MySQL user to the database, giving the MySQL user all
	 privileges.


2. Upload all the script files to your server

	This is best done with an FTP client such as FileZilla.
	Please make sure to keep the original directory structure.
	
	Note: It is not always necessary to have full permissions on all server
	configurations.  Especially if your server is running fcgi or suExec.  In
	some cases you might be able to get away with using only more restrictive
	permissions.  For example 700, 750, 770, 755 for directories or for files
	600, 640, 660, 644.
	
	More info: 
	http://en.wikipedia.org/wiki/Chmod
	http://en.wikipedia.org/wiki/Filesystem_permissions
	
	
	Set write permissions for the following directories: (chmod 777)
	
	admin/temp/
	pixels/
	
	(Reason for the permissions: The script needs to be able to create images in
	the admin/temp/ and pixels/ directories)
	
	Set write permissions to config.php (chmod 666). The script needs to write
	to this file during installation and configuration.
	
	Tips:
	
	*nix via SSH: Change permissions from the command line using the "chmod" 
	command. eg. chmod 777 pixels
	
	*nix via FTP: Use an FTP client which allows you to set permissions. 
	eg. Right click on the file.
	
	See tutorials on www.milliondollarscript.com
	
	Windows: Make sure the web-server/php process can write to pixels/ and 
	admin/temp directories


3. Run the installation install.php script
	
	Go to http://yourwebsite.com/admin/install.php
	
	The installation script will install the database for you.
	Fill in the database details from step 1.
	
	Please delete the install.php file after successful installation.


4. Edit your configuration.
	
	Go to http://yourwebsite.com/admin/ and use 'ok' as the default password.
	
	Here is what should be configured:
	
	 a. Main Config (Configuration menu.)
	 b. Languages (Edit labels, email templates, etc using the editing tool)
	 c. Payment Modules
	 d. Currencies
	
	 After configuration, go to 'Pixel Admin' and process pixels. Click on the
	 'Process Images(s)' button to generate the images.
	
	Additionally, you can set up your pixel inventory, prices, background 
	blending, etc.
	
	The admin has many possibilities, be creative ;)


5. Test the script.
	
	If running the script for the first time, go to Admin and process the pixels
	from the 'Pixel Admin' menu. 
	
	Then look at /index.php to see if the images were
	installed correctly. Finally, sign up as a new user and make a test order.
	You can set all the payment modules to demo mode if needed.


6. Create your site / integrate the script to your site.

	Look at index.php as an example of integration.


Note for 2C0 settings

	The script requires 2CO account version 2. Also you will need to set the 
	return URLs in your 2C0 account (under the Look and Feel). 2C0 agreement 
	allows only one account per website, so it is better to have the script 
	running on the same domain that you registered your 2C0 account with!


Note of PayPal settings

	To get the Instant Payment Notifications to work, ensure that 
	IPN is enabled for your PayPal account. Log into your paypal account and 
	click on 'Profile' -> 'Instant Payment Notification Preferences' and hit 
	Edit.  Enter the URL of your website. (Any URL will do, the script will 
	automatically tell PayPal the URL when an order is made)


-----------------------------

Knowledge Base / FAQ:

- There are over 100 articles in the knowledge base covering many topics.

https://milliondollarscript.com/category/faq/

-----------------------------


Tips for Use
##########################################################

> How to order without paying via Paypal / 2C0

1. Select some pixels to place them on the order. 
2. Confirm the order
3. Go into the admin/ and complete the order from the 'Orders Waiting' menu

Also, you can set an account to be in 'privileged mode, and the users
will never need to pay for the order.

##########################################################

> Sometimes people make orders but do not pay. What should I do?

The system keeps the pixels reserved for the client. If the client does not
complete the payment on time, you can delete the order from the
Admin. Deleting the order will free the pixels up for everyone else.
Non-paid orders are automatically cancelled after a number of days
as specified in config.

The script will also automatically delete unpaid orders after specified time.
The time can be specified in Main Config.

##########################################################

> Tips for Approving pixels & Processing the grid(s).

You should process your image(s) and publish it/them every time
AFTER approving your client's links.

##########################################################

> How do I log in to another user's account??

You can use the Admin password as the master password to log
in to a user's account.

The master password is the password that you have set in the config.
To log in, enter the users' username, and enter the master password.

##########################################################

> Can I edit the images for the blocks?

There are three different sets of images for blocks.
They are separated into the following directories.
You are welcome to edit them, but do not delete them - they
are needed by the script.


users/

block.png - available block
not_for_sale_block.png - Not for sale (white)
ordered_block.png - block on a confirmed order (orange)
reserved_block.png - reserved, but not confirmed yet. (yellow)
selected_block.png - Selected by user (green)
sold_block.png - completed order (red)

admin/

block.png - available block, used for preview
not_for_sale_block.png - the color of N.F.S. block for preview (green)
sold_block.png - sold block used for preview (red)

admin/temp/

block.png - available block, used for final processed grid  (white)
not_for_sale_block.png - N.F.S. block used for final processed grid (white)


##########################################################

> Having Problems generating the image?

99% of problems are due to the following:

1. Incorrect path and URL settings in Main Config. Please edit this carefully
and follow the recommendations

2. Permission problems. This is a common problem. Please check your permission 
settings.

3. Your PHP is too old. Check to make sure you have at least 5.2.x

4. The files were not uploaded with the original directory structure in tact.

5. Have you approved the pixels? Pixels need to be approved before they are
processed.

6. Don't delete any files from the admin/temp directory

######################################################

> I change header.php and footer.php, but there is no effect. Why?

The files forgot.php, signup.php, validate.php and the login form 
(login_functions.php) do not use the header and footer files because they are
external from the 'main' application.  Please modify the header tags directly in
these files. Note that the css file still affects these files.

######################################################

> I am getting 'page not found' error when i'm trying to process the pixels.

The cause is incorrectly compiled PHP on the server which is causing PHP to
crash when using the GD library bundled with PHP.
 
The problem is described here: http://bugs.php.net/bug.php?id=29568
See the 2nd last comment: 

 "pdflib uses a bundled version of png (old version), which makes php
 (using newer png) crash. Either remove --with-pdf when compiling PHP, or
 compile pdflib with "--with-pnglib --with-zlib" to make pdflib use the
 newer png version."


Please try to upgarde your PHP, or disable pdflib from PHP.

######################################################

> I am getting an error: "Fatal error: Call to undefined function imagecreate()"

If you are hosting on Windows then it means that you do not have the GD library 
turned on. 
 
Please see PHP the documentation to learn how to enable extensions in Windows.

Roughly, here is what you need to do:
 
1. Copy the file php_gd2.dll from your PHP's ext/ directory to where you have
your php extensions. (No need to copy if php_gd2.dll is in the extension dir
already. Look in php.ini for 'extension_dir' directive to find what is your
current extension directory)
 
2. Modify your php.ini and change the following line:
 
;extension=php_gd2.dll
 
to:
 
extension=php_gd2.dll
 
3. Restart your web server.

######################################################

> I would like to approve the links one at a time.

This is not possible.

Links are grouped by advertiser, and all links for that advertiser need
to be approved in one go. This makes things simple and faster to administer.

Why are the links grouped?

If you approve each link individually then the images might end up 
looking like Swiss cheese. This is because an image can have different 
links, and if you don't approve some, then you will end up having holes in the 
image.  Grouping them eliminates this problem.

########################################################

> My background image too big and I cannot upload it. What can I do?

PHP may limit upload file size to 2MB by default and will not let you to
upload a background image via the admin.

You can upload the image by FTP.

Upload the file to the admin/temp directory

Then, rename the file to background1.png (The number 1 is to identify that the 
background is to be used with the first grid. Rename to background2.png if you 
want to identify with the 2nd grid and so on)

You will be able to see the background in the Admin after it is uploaded.

########################################################

> My hosting company does not allow me to set chmod 777 permissions

Does the script show any problem, saying that it cannot write to a directory
or file? If it does not, then everything is OK.

If it cannot write to a directory or file and you can only set to 666, then 
try to do this:

Create empty files with the following names:

in admin/temp/

temp1.png
temp1.jpg
temp1.gif
background1.png

in pixels/

main1.png
main1.jpg
main1.gif

Then give chmod 666 to all the files above.

These files can be just plain text files. For example upload myfile.txt and 
rename to temp1.png, upload myfile.txt and rename to temp1.jpg, and so on...

The above example creates the files for grid 1. If you have a 2nd grid, replace
the 1 with a 2, and so on..

	Note: It is not always necessary to have full permissions on all server
	configurations.  Especially if your server is running fcgi or suExec.  In
	some cases you might be able to get away with using only more restrictive
	permissions.  For example 700, 750, 770, 755 for directories or for files
	600, 640, 660, 644.
	
	More info: 
	http://en.wikipedia.org/wiki/Chmod
	http://en.wikipedia.org/wiki/Filesystem_permissions

###################################################

> I get error SQL when registering a new user

Note: As of version 2.1 MySQL 5.7 is recommended.

This problem can also be solved by opening the file 'my.ini' in your MySQL 
install folder - the config file. And commenting out the command near the 
top that turns on the 'strict' mode.

>> -gl

####################################################

> I get errors that look like this: 
> Fatal error: Allowed memory size of 8388608 bytes exhausted
> (tried to allocate 3000 bytes)

Note: As of version 2.1 the default memory limit is set to 128M.

It looks like your PHP memory limit is set to 8MB and the script needs more. 
Please contact your hosting service to increase the limit.
Increasing to 12MB or more usually helps.

If you are running apache, you can try this:

Create a .htaccess file in your web directory (or modify if the file is 
there) and add the following line to the file:

php_value memory_limit 16M



If your hosting accound does not support .htaccess, add the following
line to include/functions.php file below the first <?php line:

ini_set('memory_limit', '12M');


If that does not work, then how big is your grid??
The script can get very thirsty if there are more than 10,000 blocks
per grid (1 million pixels) so check if you didn't set it to something
really big! Or...you may have a really big background image that's 
causing the trouble.

####################################################

> When I tried running the installation install.php script, 
> I got these notices. I'm unable to install database
>
> Notice: Use of undefined constant action - assumed 'action' 
> in d:\accounts\millionpix\admin\install.php on line 3


It looks like you have Notices turned on in your PHP php.ini file. 
This occurs when you first install php on your server, and the notices are 
turned on by default.

Notices should always be turned off in production environments.

Please edit your php.ini file and make sure that the error_reporting directive
is set to:

error_reporting  =  E_ALL & ~E_NOTICE & ~E_STRICT

Please make sure to restart your web server after editing the php.ini file.

###################################################

UPGRADES

Please check the website for updates.  http://milliondollarscript.com

Notices and instructions will be posted on the site when updates are released.
 
Unless otherwise stated, it is not required for you to upgrade.

If you want to keep your script upgradeable, please limit your customization of 
the script to only the following files: main.css, labels.php, header.php, 
footer.php and config.php
 
If you have made significant customizations, use a program such as WinDiff or
the very excellent WinMerge http://winmerge.sourceforge.net/
to help you with merging the new files.

######################################################################


/**
 * @package		mds
 * @copyright	(C) Copyright 2020 Ryan Rhode, All rights reserved.
 * @author		Ryan Rhode, ryan@milliondollarscript.com
 * @license		This program is free software; you can redistribute it and/or modify
 *		it under the terms of the GNU General Public License as published by
 *		the Free Software Foundation; either version 3 of the License, or
 *		(at your option) any later version.
 *
 *		This program is distributed in the hope that it will be useful,
 *		but WITHOUT ANY WARRANTY; without even the implied warranty of
 *		MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *		GNU General Public License for more details.
 *
 *		You should have received a copy of the GNU General Public License along
 *		with this program;  If not, see http://www.gnu.org/licenses/gpl-3.0.html.
 *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *
 *		Million Dollar Script
 *		A pixel script for selling pixels on your website.
 *
 *		For instructions see README.txt
 *
 *		Visit our website for FAQs, documentation, a list team members,
 *		to post any bugs or feature requests, and a community forum:
 * 		https://milliondollarscript.com/
 *
 */

Acknowledgments

PHPMailer
https://github.com/PHPMailer/PHPMailer

AREA Map code contributed by Martin Diekhoff
http://www.onecentads.com/

Imagine - Image manipulation library for PHP 5.3 inspired by Python's PIL and other image libraries.
https://github.com/avalanche123/Imagine

HTMLPurifier - HTML filtering solution
http://htmlpurifier.org/

jQuery JavaScript library
https://jquery.com/

jQuery Form Plugin
https://github.com/jquery-form/form

Tippy.js - Tooltip, popover, dropdown, and menu library
https://github.com/atomiks/tippyjs

Popper - Tooltip & Popover Positioning Engine (used by Tippy.js)
https://popper.js.org/

Image-Map - Responsive, dynamic image maps.
https://github.com/clarketm/image-map

Image Scale - Scale images to fit or fill any target container via two simple properties: scale and align
https://github.com/gestixi/image-scale

BrowserStack - App & Browser Testing Made Easy
https://www.browserstack.com/


And of course anything else I might be missing.

Thank you to everyone else who helps out!


CHANGELOG
<?php
/**
 * @package		mds
 * @copyright	(C) Copyright 2020 Ryan Rhode, All rights reserved.
 * @author		Ryan Rhode, ryan@milliondollarscript.com
 * @license		This program is free software; you can redistribute it and/or modify
 *		it under the terms of the GNU General Public License as published by
 *		the Free Software Foundation; either version 3 of the License, or
 *		(at your option) any later version.
 *
 *		This program is distributed in the hope that it will be useful,
 *		but WITHOUT ANY WARRANTY; without even the implied warranty of
 *		MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *		GNU General Public License for more details.
 *
 *		You should have received a copy of the GNU General Public License along
 *		with this program;  If not, see http://www.gnu.org/licenses/gpl-3.0.html.
 *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *
 *		Million Dollar Script
 *		A pixel script for selling pixels on your website.
 *
 *		For instructions see README.txt
 *
 *		Visit our website for FAQs, documentation, a list team members,
 *		to post any bugs or feature requests, and a community forum:
 * 		https://milliondollarscript.com/
 *
 */
?>

MDS 2.1 alpha

- SECURITY RELEASE - UPGRADED STRONGLY RECOMMENDED
- Please read UPGRADE.txt for upgrade instructions
- Project transferred to new project coordinator: Ryan Rhode (ryan@milliondollarscript.com)
- Fixed various SQL errors
- added HTMLPurifier to library - still has to be implemented system-wide
- added variable filters for significant security improvements
- converted UTF-8 and XML Transitional 
- changed FCKEditor to CKEditor
- removed overlib
- many other miscellaneous fixes / enhancements
- see here for a list of additional changes: https://gitlab.com/MillionDollarScript/MillionDollarScript

MDS 2.0.15

- License changed to GPL
- Copyright transferred form Jamit Software to 
Adam Malinowski (adam@milliondollarscript.com)

MDS 2.0.14

- fix max orders bug
- fix image upload bug if Safe Mode is on
- fix bug with package price not updated
- fixed 'write your ad' button to be enabled when 'Edit Order'is pressed
- Fix 'Edit Order' button


MDS 2.0.13

- delete temp order when a new pixel image gets uploaded
- adjustements to the javascript in users/select_pixels.php
- Admin Order's Map: increased height
- Admin Process Pixels: expired orders publish status set to 'N'
- users/ Logout: Logout link does not show if logged in
- Admin: Ad List Sql Query fix
- Top Advertisers Sql Query Fix


MDS 2.0.11


- Fix problem with category field types.
- Fix problem with deleting Advertiser's account form the admin.

===================================================

Million Dollar Script version 1.7.8 - 1.7.9
Released April 28, 2006 

This version contains some minor bug fixes reported by users.
There are no major features in this version.

You can upgrade to this version from any older version. No sequence is
necessary. The latest version will update your database automatically
to the latest version. Your tables and data will be preserved.
It is always a good idea to backup your data before the upgrade.

See below for upgrade instructions



#######################################################
CHANGES 

Version 1.7.8 & 1.7.9:

- Maintanance release version, includes minor fixes to be compatible with more
servers & minor reported bugs fixed.

Since Version 1.7.7:

- Fixed SMTP email bug, SMTP email test & seperated SMTP server form POP
- POP before SMTP setting
- Localized some strings in the payment module
- minor bug fixes.

Since Version 1.7.6:

- Seperated POP server & SMTP setting in Main Config
- Added a Pop-before-smtp option (beta)
- Updated Authorize.net module

Since Version 1.7.5:

- Added ccAvenue, e-gold, moneybookers and Authorize.net payment modules
- Added email alert to Admin when pixels are uploaded, admin can approve pixels and
process the image by email
- Minor bug fixes all round.

Since Version 1.7.4:

- Fixed bug with displaying confirmed order when using packages.
- Fixed bug with unlimited max orders in packages. 

Since Version 1.7.3:

- Fixed bug with displaying 'days to expire' of packages
- Fixed bug with click.php
- Fixed bug in install.php

Since Version 1.7.2:

- Fixed mouseover bugs
- Addred transition effects to mouseover
- Fixed bug with selection image not showing as GIF, when GIF was selected


Since Version 1.6.5:


- Optimized HTML, saving up to 90% bandwidth in a best case scenario. Grids will load much faster - they will scream!
- Packages: Prices options can be grouped in a package, the advertiser will be able to select the best price option from the package.
- Enhanced click tracking & click reports
- Click on available pixels will take the user to a Admin-defined URL
- Improvements to the AJAX code 
- Improvements to Approve URL



admin/nfs.php
<?php
/**
 * @package       mds
 * @copyright     (C) Copyright 2020 Ryan Rhode, All rights reserved.
 * @author        Ryan Rhode, ryan@milliondollarscript.com
 * @version       2020.05.08 17:42:17 EDT
 * @license       This program is free software; you can redistribute it and/or modify
 *        it under the terms of the GNU General Public License as published by
 *        the Free Software Foundation; either version 3 of the License, or
 *        (at your option) any later version.
 *
 *        This program is distributed in the hope that it will be useful,
 *        but WITHOUT ANY WARRANTY; without even the implied warranty of
 *        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *        GNU General Public License for more details.
 *
 *        You should have received a copy of the GNU General Public License along
 *        with this program;  If not, see http://www.gnu.org/licenses/gpl-3.0.html.
 *
 *  * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *
 *        Million Dollar Script
 *        A pixel script for selling pixels on your website.
 *
 *        For instructions see README.txt
 *
 *        Visit our website for FAQs, documentation, a list team members,
 *        to post any bugs or feature requests, and a community forum:
 *        https://milliondollarscript.com/
 *
 */

require_once __DIR__ . "/../include/init.php";
require( 'admin_common.php' );

ini_set( 'max_execution_time', 10000 );
ini_set( 'max_input_vars', 10002 );

$BID = $f2->bid();

$banner_data = load_banner_constants( $BID );

$imagine = new Imagine\Gd\Imagine();

// load blocks
$block_size  = new Imagine\Image\Box( $banner_data['BLK_WIDTH'], $banner_data['BLK_HEIGHT'] );
$palette     = new Imagine\Image\Palette\RGB();
$color       = $palette->color( '#000', 0 );
$zero_point  = new Imagine\Image\Point( 0, 0 );
$blank_block = $imagine->create( $block_size, $color );

$default_nfs_block = $blank_block->copy();
$tmp_block         = $imagine->load( $banner_data['USR_NFS_BLOCK'] );
$tmp_block->resize( $block_size );
$default_nfs_block->paste( $tmp_block, $zero_point );
$data = base64_encode( $default_nfs_block->get( "png", array( 'png_compression_level' => 9 ) ) );

if ( $_REQUEST['action'] == 'save' ) {
	//$sql = "delete from blocks where status='nfs' AND banner_id=$BID ";
	//mysqli_query($GLOBALS['connection'], $sql) or die (mysqli_error($GLOBALS['connection']).$sql);

	if ( isset( $_REQUEST['addnfs'] ) && ( ! empty( $_REQUEST['addnfs'] || $_REQUEST['addnfs'] == 0 ) ) ) {
		$addnfs = explode( "~", $_REQUEST['addnfs'] );
	} else {
		unset( $addnfs );
	}

	if ( isset( $_REQUEST['remnfs'] ) && ( ! empty( $_REQUEST['remnfs'] ) || $_REQUEST['remnfs'] == 0 ) ) {
		$remnfs = explode( "~", $_REQUEST['remnfs'] );
	} else {
		unset( $remnfs );
	}

	$cell = $x = $y = 0;
	for ( $i = 0; $i < $banner_data['G_HEIGHT']; $i ++ ) {
		$x = 0;
		for ( $j = 0; $j < $banner_data['G_WIDTH']; $j ++ ) {
			if ( isset( $addnfs ) && in_array( $cell, $addnfs ) ) {
				$sql = "REPLACE INTO blocks (block_id, status, x, y, image_data, banner_id, click_count) VALUES ({$cell}, 'nfs', {$x}, {$y}, '{$data}', {$BID}, 0)";
				mysqli_query( $GLOBALS['connection'], $sql ) or die( mysqli_error( $GLOBALS['connection'] ) . $sql );
			}
			if ( isset( $remnfs ) && in_array( $cell, $remnfs ) ) {
				$sql = "DELETE FROM blocks WHERE status='nfs' AND banner_id={$BID} AND block_id={$cell}";
				mysqli_query( $GLOBALS['connection'], $sql ) or die( mysqli_error( $GLOBALS['connection'] ) . $sql );
			}
			$x = $x + $banner_data['BLK_WIDTH'];
			$cell ++;
		}
		$y = $y + $banner_data['BLK_HEIGHT'];
	}
	echo "Success!";
	exit();
} else if ( $_REQUEST['action'] == 'reset' ) {
	$sql = "DELETE FROM blocks WHERE status='nfs' AND banner_id=$BID";
	mysqli_query( $GLOBALS['connection'], $sql ) or die( mysqli_error( $GLOBALS['connection'] ) . $sql );
	echo "Success!";
	exit();
}
?>
<script type="text/javascript">
	$(function () {
		var addnfs = [];
		var remnfs = [];

		function processBlock(block) {
			if (block.hasClass("nfs")) {
				block.removeClass("nfs").addClass("free");
				var blockid = block.attr("data-block");
				remnfs.push(blockid);
				var index = addnfs.indexOf(blockid);
				if (index != -1) {
					addnfs.splice(index, 1);
				}
			} else if (block.hasClass("free")) {
				block.removeClass("free").addClass("nfs");
				var blockid = block.attr("data-block");
				addnfs.push(blockid);
				var index = remnfs.indexOf(blockid);
				if (index != -1) {
					remnfs.splice(index, 1);
				}
			} else if (!block.hasClass("free")) {
				block.removeClass("ui-selected");
			}
		}

		$('.grid').selectable({
			delay: 75,
			distance: <?php echo $banner_data['BLK_WIDTH']; ?>,
			autoRefresh: false,
			filter: 'span',
			stop: function () {
				$(".ui-selected", this).each(function () {
					processBlock($(this));
				});
				$(this).selectable("refresh");
			}
		});

		$(".block").click(function () {
			processBlock($(this));
		});

		$('.save').click(function (e) {
			e.preventDefault();
			e.stopPropagation();
			$(this).after('<img class="loading" width="16" height="16" src="../images/ajax-loader.gif" alt="Loading..." />');
			$('.save').prop('disabled', true);
			$('.reset').prop('disabled', true);

			var posting = $.post("nfs.php", {
				BID: <?php echo $BID; ?>,
				action: "save",
				addnfs: addnfs.join('~'),
				remnfs: remnfs.join('~')
			});

			posting.done(function (data) {
				$('.loading').hide(function () {
					$(this).remove();
					$('<span class="message">' + data + '</span>').insertAfter('.save').fadeOut(10000, function () {
						$(this).remove();
					});
				});
				$('.save').prop('disabled', false);
				$('.reset').prop('disabled', false);
			});
		});

		$('.reset').click(function (e) {
			e.preventDefault();
			e.stopPropagation();
			var result = confirm('This will unselect all NFS blocks. Are you sure you want to do this?');
			if (result) {
				$('.save').prop('disabled', true);
				$('.reset').prop('disabled', true);
				$(this).after('<img class="loading" width="16" height="16" src="../images/ajax-loader.gif" alt="Loading..." />');
				var posting = $.post("nfs.php", {
					BID: <?php echo $BID; ?>,
					action: "reset"
				});
				posting.done(function (data) {
					$('.loading').hide(function () {
						$(this).remove();
						$('<span class="message">' + data + '</span>').insertAfter('.reset').fadeOut(10000, function () {
							$(this).remove();
						});
					});
					$('.save').prop('disabled', false);
					$('.reset').prop('disabled', false);
					$(".block.nfs").removeClass("ui-selected").removeClass("nfs").addClass("free");
				});
			}
		});
	});
</script>
<style type="text/css">
    <?php
	$grid_background = "";
		if(file_exists(BASE_PATH . "/" . BANNER_DIR . "main$BID.png")) {
			$grid_background = 'background: url("../' . BANNER_DIR . 'main' . $BID . '.png") no-repeat center center;background-size:contain;';
	}
	?>
    .grid {
    <?php echo $grid_background; ?> z-index: 0;
        width: <?php echo $banner_data['G_WIDTH']*$banner_data['BLK_WIDTH']; ?>px;
        height: <?php echo $banner_data['G_HEIGHT']*$banner_data['BLK_HEIGHT']; ?>px;
    }

    .block_row {
        clear: both;
        display: block;
    }

    .block {
        white-space: nowrap;
        width: <?php echo $banner_data['BLK_WIDTH']; ?>px;
        height: <?php echo $banner_data['BLK_HEIGHT']; ?>px;
        float: left;
        opacity: 0.5;
        filter: alpha(opacity=50);
    }

    .sold {
        background: url("../images/sold_block.png") no-repeat;
    }

    .reserved {
        background: url("../images/reserved_block.png") no-repeat;
    }

    .nfs {
        background: url('data:image/png;base64,<?php echo $data; ?>') no-repeat;
        cursor: pointer;
    }

    .ordered {
        background: url("../images/ordered_block.png") no-repeat;
    }

    .ordered {
        background: url("../images/ordered_block.png") no-repeat;
    }

    .onorder {
        background: url("../images/not_for_sale_block.png") no-repeat;
    }

    .free {
        background: url("../images/block.png") no-repeat;
        cursor: pointer;
    }

    .grid .ui-selecting {
        background: #FECA40;
    }
</style>

<div class="outer_box">
    <p>
        Here you can mark blocks to be not for sale. You can drag to select an area. Click 'Save' when done.
    </p>
    (Note: If you have a background image, the image is blended in using the browser's built-in filter - your alpha channel is ignored on this page)
    <hr>
	<?php
	$sql = "Select * from banners ";
	$res = mysqli_query( $GLOBALS['connection'], $sql );
	?>
    <form name="bidselect" method="post" action="nfs.php">
        <label>
            Select grid:
            <select name="BID" onchange="mds_submit(this)">
                <option></option>
				<?php
				while ( $row = mysqli_fetch_array( $res ) ) {
					if ( ( $row['banner_id'] == $BID ) && ( $BID != 'all' ) ) {
						$sel = 'selected';
					} else {
						$sel = '';
					}
					echo '
                <option
                ' . $sel . ' value=' . $row['banner_id'] . '>' . $row['name'] . '</option>';
				}
				?>
            </select>
        </label>
    </form>
    <hr>
	<?php
	if ( $BID != '' ) {
	$sql    = "show columns from blocks ";
	$result = mysqli_query( $GLOBALS['connection'], $sql );
	while ( $row = mysqli_fetch_array( $result ) ) {
		if ( $row['Field'] == 'status' ) {
			if ( strpos( $row['Type'], 'nfs' ) == 0 ) {
				$sql = "ALTER TABLE `blocks` CHANGE `status` `status` SET( 'reserved', 'sold', 'free', 'ordered', 'nfs' ) NOT NULL ";
				mysqli_query( $GLOBALS['connection'], $sql ) or die ( "<p><b>CANNOT UPGRADE YOUR DATABASE!<br>Please run the follwoing query manually from PhpMyAdmin:</b><br>$sql<br>" );
			}
		}
	}
	$sql = "select block_id, status, user_id FROM blocks WHERE banner_id=" . intval( $BID );
	$result = mysqli_query( $GLOBALS['connection'], $sql ) or die ( mysqli_error( $GLOBALS['connection'] ) . $sql );
	while ( $row = mysqli_fetch_array( $result ) ) {
		$blocks[ $row["block_id"] ] = $row['status'];
	}
	?>
    <div class="container">
        <input class="save" type="submit" value='Save Not for Sale'/>
        <input class="reset" type="submit" value='Reset'/>
        <div class="grid">
			<?php
			$cell = "0";
			for ( $i = 0; $i < $banner_data['G_HEIGHT']; $i ++ ) {
				echo "<div class='block_row'>";
				for ( $j = 0; $j < $banner_data['G_WIDTH']; $j ++ ) {
					switch ( $blocks[ $cell ] ) {
						case 'sold':
							echo '<span class="block sold" data-block="' . $cell . '"></span>';
							break;
						case 'reserved':
							echo '<span class="block reserved" data-block="' . $cell . '"></span>';
							break;
						case 'nfs':
							echo '<span class="block nfs" data-block="' . $cell . '"></span>';
							break;
						case 'ordered':
							echo '<span class="block ordered" data-block="' . $cell . '"></span>';
							break;
						case 'onorder':
							echo '<span class="block onorder" data-block="' . $cell . '"></span>';
							break;
						case 'free':
						case '':
							echo '<span class="block free" data-block="' . $cell . '"></span>';
					}
					$cell ++;
				}
				echo '</div>
				';
			}
			?>
        </div>
        <input class="save" type="submit" value='Save Not for Sale'/>
    </div>
</div>
<?php } ?>
